mungegithub/mungers/milestone-maintainer.go

```diff
@@ -0,0 +1,461 @@
+
+	// Warn about milestone-required labels every day
+	milestoneWarningPeriod = 1 * day
+
+	priorityCriticalUrgent = "priority/critical-urgent"
+
+	commentDetail = `<details>
+Additional instructions available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/release/issues.md)
+</details>`
+)
+
+var (
+	kindMap = map[string]string{
+		"kind/bug":     "Fixes a bug discovered during the current release.",
+		"kind/feature": "New functionality.",
+		"kind/cleanup": "Adding tests, refactoring, fixing old bugs.",
+	}
+
+	priorityMap = map[string]string{
+		priorityCriticalUrgent:        "Never automatically move out of a release milestone; continually escalate to contributor and SIG through all available channels.",
+		"priority/important-soon":     "Escalate to the issue owners and SIG owner; move out of milestone after several unsuccessful escalation attempts.",
+		"priority/important-longterm": "Escalate to the issue owners; move out of the milestone after 1 attempt.",
+	}
+
+	milestoneRemovedLabel          = strings.ToLower(strings.Replace(milestoneRemoved, " ", "-", -1))
+	milestoneLabelsIncompleteLabel = strings.ToLower(strings.Replace(milestoneLabelsIncomplete, " ", "-", -1))
+	milestoneLabelsCompleteLabel   = strings.ToLower(strings.Replace(milestoneLabelsComplete, " ", "-", -1))
+	milestoneLabels                = []string{milestoneRemovedLabel, milestoneLabelsCompleteLabel, milestoneLabelsIncompleteLabel}
+)
+
+// milestoneNotification configures the active notification for the munger
+type milestoneNotification struct {
+	deleteExisting bool
+	description    string
+	message        string
+}
+
+// MilestoneMaintainer enforces the process for sheperding issues into the release.
+type MilestoneMaintainer struct {
+	botName  string
+	features *features.Features
+
+	activeMilestone string
+}
+
+func init() {
+	RegisterMungerOrDie(&MilestoneMaintainer{})
+}
+
+// Name is the name usable in --pr-mungers
+func (m *MilestoneMaintainer) Name() string { return "milestone-maintainer" }
+
+// RequiredFeatures is a slice of 'features' that must be provided
+func (m *MilestoneMaintainer) RequiredFeatures() []string { return []string{} }
+
+// Initialize will initialize the munger
+func (m *MilestoneMaintainer) Initialize(config *github.Config, features *features.Features) error {
+	if len(m.activeMilestone) == 0 {
+		return errors.New("active-milestone must be supplied")
+	}
+	m.botName = config.BotName
+	m.features = features
+	return nil
+}
+
+// EachLoop is called at the start of every munge loop
+func (m *MilestoneMaintainer) EachLoop() error { return nil }
+
+// RegisterOptions registers options for this munger; returns any that require a restart when changed.
+func (m *MilestoneMaintainer) RegisterOptions(opts *options.Options) sets.String {
+	opts.RegisterString(&m.activeMilestone, "active-milestone", "", "The active milestone that this munger will maintain issues for.")
+	return sets.NewString("active-milestone")
```

Comment:
The `active-milestone` option can be reloaded without restarting mungegithub. It just needs to be revalidated (in this case check that it has length > 0). So `RegisterOptions` should return `nil` and an option update callback should be defined and registered with `opts` from within this function. The callback can check if the option changed, and if so it can check if the option is still valid and throw an error if not. See the [submit-queue](https://github.com/kubernetes/test-infra/blob/master/mungegithub/mungers/submit-queue.go#L593) for an example. This will allow the `active-milestone` option to be updated in the configmap without restarting the mungegithub instance.